File: cloud_extel/cloud_extel/report/chart_of_account_report/chart_of_account_report.js
// Copyright (c) 2022, Frappe and contributors
// For license information, please see license.txt
/* eslint-disable */

frappe.query_reports["Chart of Account Report"] = {
	"filters": [
		{
			"fieldname": "company",
			"label": __("Company"),
			"fieldtype": "Link",
			"options": "Company",
			"default": frappe.defaults.get_user_default("Company"),
			"reqd": 1
		},
		{
			"fieldname": "name",
			"label": __("Account"),
			"fieldtype": "Link",
			"options": "Account",
			"get_query": function() {
				let company = frappe.query_report.get_filter_value('company');
				return {
					filters: {
						company: company
					}
				}
			}
		},
	],
	"tree": true,
	"name_field": "account",
	"parent_field": "parent_account",
	"initial_depth": 3,
	"formatter": function(value, row, column, data, default_formatter) {
		if (column.fieldname == "amount" && data) {
			let dr_or_cr = data.amount > 0 ? "Dr": "Cr";
			value = format_currency(Math.abs(data.amount), data.account_currency) + ' ' + dr_or_cr;
		}

		return value;
	},
}

File: cloud_extel/cloud_extel/report/chart_of_account_report/chart_of_account_report.py
# Copyright (c) 2022, Frappe and contributors
# For license information, please see license.txt

import frappe
from frappe import _
from erpnext.accounts.report.financial_statements import filter_accounts
from erpnext.accounts.utils import get_account_balances

def execute(filters=None):
	columns = get_columns(filters)
	data = get_data(filters)

	return columns, data

def get_columns(filters):
	return [
		{
			"fieldname": "value",
			"label": _("Account"),
			"fieldtype": "Link",
			"options": "Account",
			"width": 300,
		},
		{
			"fieldname": "amount",
			"label": _("Amount"),
			"fieldtype": "Data",
			"width": 300,
		},
		{
			"fieldname": "account_currency",
			"label": _("Currency"),
			"fieldtype": "Link",
			"options": "Currency",
			"hidden": 1,
		},
	]

def get_data(filters):
	report_filters = {}
	for field in ["name", "company"]:
		if filters.get(field):
			report_filters[field] = filters.get(field)

	accounts = frappe.get_all(
		"Account",
		filters = report_filters,
		fields = ["name", "name as value", "account_number",
			"parent_account", "account_name", "account_currency"]
	)

	accounts = filter_accounts(accounts)[0]
	accounts = get_account_balances(accounts, filters.get("company"))

	for account in accounts:
		account["amount"] = account.get("balance_in_account_currency") or account.get("balance")

	return accounts
